### PR TITLE
Implement global security advisories API

### PR DESCRIPTION
**Pull request by cpanato**
- implement global security advisories API

Fixes: https://github.com/google/go-github/issues/2851

cc @gmlewis

### CODE REVIEW

File: github/security_advisories.go
Comment (gmlewis) [inline github/security_advisories.go:69]: Every field with the `omitempty` tag should either be a pointer or a slice.  Please change them.  Line 77, for example, is fine as-it-is because it is a slice of strings.  A slice to structs, however, would need to be a slice to struct pointers.

Also, since I'm here, and this is a nit, please change `//If` to `// If`.

File: github/security_advisories.go
Comment (gmlewis) [inline github/security_advisories.go:110]:
```suggestion
	Identifiers           []*Identifiers     `json:"identifiers,omitempty"`
```

File: github/security_advisories.go
Comment (gmlewis) [inline github/security_advisories.go:120]:
```suggestion
	Vulnerabilities       []*AdvisoryVulnerability `json:"vulnerabilities,omitempty"`
	CVSS                  *CVSS              `json:"cvss,omitempty"`
	CWEs                  []*CWEs            `json:"cwes,omitempty"`
	Credits               []*UserCredit         `json:"credits,omitempty"`
```

**Review by gmlewis — CHANGES_REQUESTED** [review]
Thanks, @cpanato. I'm going to stop my code review early here and ask you to remove as many new structs that you've added as possible, and to instead reuse `SecurityAdvisory` and other related structs as much as possible without all the duplication.

Also, every new exported struct that you add needs a full godoc string.

File: github/security_advisories.go
Comment (cpanato) [reply github/security_advisories.go:69]: thank you for your feedback!

File: github/security_advisories.go
Comment (gmlewis) [inline github/security_advisories.go:55]:
```suggestion
	GHSAID *string `url:"ghsa_id,omitempty"`
```

File: github/security_advisories.go
Comment (gmlewis) [inline github/security_advisories.go:61]:
```suggestion
	Type *string `url:"type,omitempty"`
```

File: github/security_advisories.go
Comment (gmlewis) [inline github/security_advisories.go:64]:
```suggestion
	CVEID *string `url:"cve_id,omitempty"`
```

File: github/security_advisories.go
Comment (gmlewis) [inline github/security_advisories.go:68]:
```suggestion
	Ecosystem *string `url:"ecosystem,omitempty"`
```

File: github/security_advisories.go
Comment (gmlewis) [inline github/security_advisories.go:72]:
```suggestion
	Severity *string `url:"severity,omitempty"`
```

File: github/security_advisories.go
Comment (gmlewis) [inline github/security_advisories.go:79]:
```suggestion
	IsWithdrawn *bool `url:"is_withdrawn,omitempty"`
```

File: github/security_advisories.go
Comment (gmlewis) [inline github/security_advisories.go:85]:
```suggestion
	Affects *string `url:"affects,omitempty"`
```

File: github/security_advisories.go
Comment (gmlewis) [inline github/security_advisories.go:88]:
```suggestion
	Published *string `url:"published,omitempty"`
```

File: github/security_advisories.go
Comment (gmlewis) [inline github/security_advisories.go:91]:
```suggestion
	Updated *string `url:"updated,omitempty"`
```

File: github/security_advisories.go
Comment (gmlewis) [inline github/security_advisories.go:94]:
```suggestion
	Modified *string `url:"modified,omitempty"`
```

File: github/security_advisories.go
Comment (gmlewis) [inline github/security_advisories.go:203]:
```suggestion
// ListGlobalSecurityAdvisories lists all global security advisories.
```

File: github/security_advisories.go
Comment (gmlewis) [inline github/security_advisories.go:208]:
```suggestion
func (s *SecurityAdvisoriesService) ListGlobalSecurityAdvisories(ctx context.Context, opts *ListGlobalSecurityAdvisoriesOptions) ([]*GlobalSecurityAdvisory, *Response, error) {
```

File: github/security_advisories.go
Comment (gmlewis) [inline github/security_advisories.go:229]:
```suggestion
// GetGlobalSecurityAdvisories gets a global security advisory using its GitHub Security Advisory (GHSA) identifier.
```

File: github/security_advisories.go
Comment (gmlewis) [inline github/security_advisories.go:121]:
```suggestion
	User *User `json:"user,omitempty"`
```

File: github/security_advisories.go
Comment (gmlewis) [inline github/security_advisories.go:105]:
```suggestion
	Vulnerabilities       []*AdvisoryVulnerability `json:"vulnerabilities,omitempty"`
```

File: github/security_advisories.go
Comment (gmlewis) [inline github/security_advisories.go:117]: Please delete this struct.

File: github/security_advisories.go
Comment (gmlewis) [inline github/security_advisories.go:108]: `Credit` is singular, because it is one instance of many.

```suggestion
	Credits               []*Credit         `json:"credits,omitempty"`
```

File: github/security_advisories.go
Comment (gmlewis) [inline github/security_advisories.go:120]:
```suggestion
// Credit represents the credit object for a global security advisory.
type Credit struct {
```

File: github/security_advisories_test.go
Comment (gmlewis) [inline github/security_advisories_test.go:321]:
```suggestion
func TestListGlobalSecurityAdvisories(t *testing.T) {
```

File: github/security_advisories_test.go
Comment (gmlewis) [inline github/security_advisories_test.go:500]:
```suggestion
	const methodName = "ListGlobalSecurityAdvisories"
	testBadOptions(t, methodName, func() (err error) {
		badOpts := &ListGlobalSecurityAdvisoriesOptions{CVEID: "CVE-\n-1234"}
		_, resp, err := client.SecurityAdvisories.ListGlobalSecurityAdvisories(ctx, badOpts)
		return err
	})
```

File: github/security_advisories.go
Comment (gmlewis) [reply github/security_advisories.go]: Sorry I wasn't clear before, as this request was not addressed in the last force-push.
I'll try to be more explicit this time.

Also, please do not use force-push, as it makes it more difficult for reviewers to see what changed since the last review.  We also always use squash-and-merge in this repo anyway, so you can have as many commits as you want and they will all be squashed down to a single commit when we merge.

**Review by gmlewis — CHANGES_REQUESTED** [review]

File: github/security_advisories_test.go
Comment (cpanato) [reply github/security_advisories_test.go:500]: This will return not found and not an error, cannot use this test as is, because we do not validate the options.

**Comment by cpanato** [comment]
@gmlewis thanks so much for your review and sorry about the force-push

File: github/security_advisories.go
Comment (gmlewis) [reply github/security_advisories.go:105]: Ah, I see.  Thanks for the explanation.

The please rename `Vulnerabilities` (plural) to `GlobalSecurityVulnerability` (singular).

File: github/security_advisories.go
Comment (gmlewis) [inline github/security_advisories.go:106]: To be consistent with this repo, please rename this field:

```suggestion
	GithubReviewedAt      *Timestamp         `json:"github_reviewed_at,omitempty"`
```

We had a discussion years ago about how to handle this situation since their trademarked name has a capital `H`, and this is how it was resolved.

File: github/security_advisories_test.go
Comment (gmlewis) [reply github/security_advisories_test.go:500]: Ah, interesting.  OK, thanks.

**Review by gmlewis — CHANGES_REQUESTED** [review]
OK, two small nits, please, then we should be ready for a second LGTM+Approval from any other contributor to this repo before merging.

File: github/security_advisories.go
Comment (cpanato) [reply github/security_advisories.go:106]: ahh, my bad

File: github/security_advisories.go
Comment (cpanato) [reply github/security_advisories.go]: thank you, done

**Comment by cpanato** [comment]
@gmlewis Thanks again for your review, PTAL

File: github/security_advisories.go
Comment (gmlewis) [inline github/security_advisories.go:111]: tiny nit:

```suggestion
// GlobalSecurityVulnerability represents a vulnerabilities for a global security advisory.
```

**Review by gmlewis — CHANGES_REQUESTED** [review]
Thank you, @cpanato !
LGTM after one tiny nit, after which we should be ready for a second LGTM+Approval from any other contributor to this repo before merging.

File: github/security_advisories.go
Comment (cpanato) [reply github/security_advisories.go]: thanks, done

Comment (gmlewis) [reply github/security_advisories.go]: WHUPS!!! My bad!!!  Sorry!

"a vulnerability" not "a vulnerabilities".

Comment (cpanato) [reply github/security_advisories.go]: and i am dumb, sorry

Comment (cpanato) [reply github/security_advisories.go]: done

**Review by gmlewis — APPROVED** [review]
Thank you, @cpanato !
LGTM.

Awaiting second LGTM+Approval from any other contributor to this repo before merging.

**Comment by cpanato** [comment]
@gmlewis checking if you know how long this will take? :)
also, after we merge this when can we have a new release? thanks again

**Comment by gmlewis** [comment]
@cpanato - we rely on volunteers to help maintain this repo. I don't know when another contributor will be available to review this PR before merging.  And yes, we should be able to make a new release soon as well.

**Comment by cpanato** [comment]
thank you @gmlewis, no worries just checking :)

**Review by valbeat — APPROVED** [review]
LGTM!

**Comment by gmlewis** [comment]
Thank you, @valbeat !
Merging.

**Comment by cpanato** [comment]
thanks for all the support

**Comment by gmlewis** [comment]
> @gmlewis checking if you know how long this will take? :) also, after we merge this when can we have a new release? thanks again

Hi @cpanato - this is now released here: https://github.com/google/go-github/releases/tag/v57.0.0
Sorry for the delay.

**Comment by cpanato** [comment]
> > @gmlewis checking if you know how long this will take? :) also, after we merge this when can we have a new release? thanks again
> 
> 
> 
> Hi @cpanato - this is now released here: https://github.com/google/go-github/releases/tag/v57.0.0
> 
> Sorry for the delay.

No sorry is needed. Thank you for all the work you do here 🎉 and thanks for the release

**Pull request merged**
